Use lowerCamelCase for the JSON API (fixes #1338)

pascalj opened the pull request:
Replace the current mix of UpperCamelCase und lowerCamelCase with consistent lowerCamelCase keys for the JSON API. Also adapt the frontend so it works with the changed API. Abbreviations are now lowercase, so `UseTLS` becomes `useTls` and so on.

As for the separator issue mentioned in #1338: the `/rest/system` endpoint already sends the `pathSeparator` that is used. That way, no Information is lost and consumers can still replace the separator if needed. I didn't make any changes regarding the separator.

**Attention**: this will break existing consumers of the API.

~~This PR depends on https://github.com/pascalj/protocol/commit/4d6a3a8831855329498c5044f4ef230bbfeb2fa9~~
I added a custom marshaller so https://github.com/pascalj/protocol/commit/4d6a3a8831855329498c5044f4ef230bbfeb2fa9 is not needed anymore.

canton7 commented:
Probably worth posting to Dev on the forums about this, if it's accepted?

AudriusButkevicius commented:
We will do that I believe.

AudriusButkevicius commented:
This an awful lot of work, thank you very much for your time.
Looks good to me, but let's wait for @calmh

calmh commented:
What Audrius says! However I would like to just one minor nit (since we are hopefully only doing this once) - lets stick to the Go convention of treating abbreviations as such. So `ID` -> `id` but `DeviceID`-> `deviceID` and `UseTLS` -> `useTLS`.

calmh commented on gui/scripts/syncthing/core/controllers/syncthingController.js:
```diff
@@ -754,12 +753,12 @@ angular.module('syncthing.core')
         };
 
-        $scope.addNewDeviceID = function (device) {
+        $scope.addNewdeviceId = function (device) {
```
This is spurious. But also the only such thing I can see which is impressive! :)

calmh commented:
@st-jenkins add to whitelist

calmh commented:
@st-jenkins run integration tests, please

calmh commented:
We are currently in the process of changing the license from GPL to MPL. If you could please drop a :+1: on #1421 to indicate that you're on board with this that would be great.

calmh commented:
Integration passes (http://build.syncthing.net/job/syncthing-pr-integration/31/).

calmh commented:
@st-jenkins test this please

calmh commented:
Aha... Mail shenanigans.

calmh commented:
@st-jenkins test this please

pascalj commented:
Thank you! I gave a :+1: on the MPL.
I changed `deviceId` to `deviceID`, `useTls` to `useTLS` and also corrected the `addNewdeviceId` error.

Since this will break the API anyway, it might be a good idea to start versioning?

calmh commented:
You thinking of adding a version integer to the /rest/system result, or something more elaborate?

pascalj commented:
Well, there are several options:
- add a version to `/rest/system` (kind of read only)
- explicit versioning of the endpoints like `/rest/v1/config...`
- more subtle with HTTP headers: "give me version x" and "this response is version y"

I like the path versioning approach because it is explicit and forces the consumer to choose a version.

But maybe there will be other occasions to introduce API namespaces/versions: I guess the API will change a few more times until 1.0 is out.

Zillode commented:
I'm in favor for ```/rest/v1/```

canton7 commented:
Consumers can currently remain compatible by ignoring case when deserializing, whereas there will have to be some very carefully-timed releases if we change the paths... 

That said, I'm also in favour of `/rest/v1` in principle. Maybe we could make `/rest/system` redirect to `/rest/v1/system` for now? Or keep the old paths alive for an iteration or two?

calmh commented:
The only real reason in having a `/rest/v1` would be to be able to have it in parallell with a `/rest/v0` which I think we are not aiming to do? Generally speaking, since this is a compatibility breaking change it should trigger a v0.11 release (or a new major if we were in 1.0). As for compatibility beyond that, we're still in beta... Although rapidly approaching the point where that's not a viable claim any more.

AudriusButkevicius commented:
I feel we just roll out v0.11 and sadly people integrating will have to do `toLower()`.

AudriusButkevicius commented:
Also, while we're at incrementing to 0.11, we should focus on getting any of the `nearly finished dangerous stuff in`, such as persistent indexes or DST or something, and spin a beta for a while.

pascalj commented:
Thanks for all your comments on versioning the API. Using the software version seems reasonable, especially since it's still in beta. I asked because I'm writing a consumer and was not sure how I can check for a compatible API.

Regarding this PR: I removed the changes to syncthing/protocol and added a custom `MarshalJSON` method for the affected struct. It should be more obvious and is not isolated from the other changes.

AudriusButkevicius commented:
Can you really define a custom marshalling methods across a package?
News to me.

pascalj commented:
You can't, but `protocol.Statistics` just gets embedded in `model.ConnectionInfo` and I implemented `MarshalJSON` for `ConnectionInfo`.

calmh commented:
Neat. Mind rebasing this on master and regenerating the assets stuff so it's mergeable again?

calmh commented:
:heart:

AudriusButkevicius commented:
:tophat:

pascalj commented:
:tada: Thanks for merging!
